### PR TITLE
ROX-14693: reconcile(create/update operation) declarative resources

### PR DESCRIPTION
**Pull request by ivan-degtiarenko**
## Description

Within this PR we create/update declarative resources found within the declarative configuration during reconciliation.
Deletion of declarative resources will be handled in a separate ticket.

The approach taken within this PR is to reconcile each resource individually. Reconciliation is done in order from referenced resources to referring resources(for example, access scope is referenced by role, so access scopes are reconciled first). That allows to avoid resolving references as it is guaranteed that in a valid configuration referenced resources will be reconciled first. 
It is possible that some state of configuration will not be valid(for example, if referenced and referring resources are located in different files), but important thing is that the configuration state in the backend will *eventually* catch on to changes within the files.

**A few other decisions made in this PR:**
1. The alternative to writing `reconcile*` functions in specific order was to introduce `protoTypeUpsertOrder` list of proto types. However, the problem was to generalize the type of reconcile function because of generic limitations in Golang.
2. `ReconciliationErrorReporter` interface decided to be made similar to `integrationHealthReporter`(every resource health and errors are treated separately). However, this is not principal part of this PR and can be changed later
3.  For auth providers `loginURL` field is filled within `registry.CreateProvider` call, `enabled` and `active` is set by default to true during transformation. There is no special handling of `basic` auth providers since declarative auth provider can't be of `basic` type.
4. Every datastore(except auth provider) got `Upsert` function. An alternative to supporting this function would be to execute `Get*` and then conditionally `Update*` or `Add*` datastore methods. There are two reasons not to follow the alternative approach: saving extra `Get*` call and not delegating datastore responsibilities to the datastore client.

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed

**Testing plan(TBE - to be executed)**:
1. Manual test(configure declarative config - see if it's being reconciled)
2. Datastore level tests

Manually testing declarative configurations:
1. Create the helm chart:
```bash
roxctl helm output central-services --image-defaults=opensource
```
2. Define the following `values.yaml` file:
```yaml
central:
  declarativeConfiguration:
    mounts:
      configMaps:
        - permission-set
        - access-scope
        - auth-provider
        - role
customize:
  envVars:
    ROX_DECLARATIVE_CONFIGURATION: true
```
3. Install the central services chart:
```bash
helm install stackrox-central-services -n stackrox -f ./values.yaml ./stackrox-central-services-chart
```
4. Verify that central is running and the `ROX_DECLARATIVE_CONFIGURATION` variable is set. The mount points for declarative configuration should exist as well.
5. Create the first config map containing a permission set:
```bash
kubectl apply -f - <<EOF
apiVersion: v1
kind: ConfigMap
metadata:
  name: permission-set
  namespace: stackrox
data:
  permission-set: |
    name: hello-world
    description: the first declaratively created permission set
    resources:
    - resource: Integration
      access: READ_ACCESS
    - resource: Administration
      access: READ_ACCESS
    - resource: Access
      access: READ_ACCESS
EOF
```
6. Verify within the a) logs of central that no error occurred during reconciliation b) within the UI / API the permission set `hello-world` exists.
7. Create the second config map containing an access scope:
```bash
kubectl apply -f - <<EOF
apiVersion: v1
kind: ConfigMap
metadata:
  name: access-scope
  namespace: stackrox
data:
  access-scope: |
    name: hello-world
    description: the first declaratively created access scope
    rules:
      included:
      - cluster: remote
EOF
```
8. Verify within the a) logs of central that no error occurred during reconciliation b) within the UI / API the access scope `hello-world` exists.
9. Create the third config map containing a role:
```bash
kubectl apply -f - <<EOF
apiVersion: v1
kind: ConfigMap
metadata:
  name: role
  namespace: stackrox
data:
  role: |
    name: hello-world
    description: the first declaratively created role
    permissionSet: hello-world
    accessScope: hello-world
EOF
```
10. Verify within the a) logs of central that no error occurred during reconciliation b) within the UI / API the role`hello-world` exists.
11. Create the fourth config map containing an auth provider:
```bash
apiVersion: v1
kind: ConfigMap
metadata:
  name: auth-provider
  namespace: stackrox
data:
  auth-provider: |
    name: hello-world
    minimumRole: "None"
    uiEndpoint: localhost:8000
    groups:
    - key: "email"
      value: "<redacted>"
      role: "Admin"
    oidc:
      issuer: <redacted>
      mode: fragment
      clientID: <redacted>

EOF
```
12. Verify within the a) logs of central that no error occurred during reconciliation b) within the UI / API the auth provider`hello-world` exists and you are able to successfully login, and group mapping works.

### CODE REVIEW

File: central/declarativeconfig/manager_impl.go
Comment (dhaus67) [inline central/declarativeconfig/manager_impl.go:192]: We should definitely add tests for upserting.

File: central/declarativeconfig/manager_impl.go
Comment (dhaus67) [inline central/declarativeconfig/manager_impl.go:37]: Albeit its just a WIP and to be changed - let's move this definition to `singleton.go` for now.

File: central/group/datastore/datastore_impl.go
Comment (dhaus67) [inline central/group/datastore/datastore_impl.go:50]: nit: move the function to the helpers function in the section below.

File: central/group/datastore/datastore_impl.go
Comment (dhaus67) [inline central/group/datastore/datastore_impl.go:51]: Shouldn't we only require a ID _iff_ the group didn't exist beforehand? Or do we want to proclaim that for `Upsert`, the ID must be set at all times?

File: central/group/datastore/datastore_impl.go
Comment (dhaus67) [inline central/group/datastore/datastore_impl.go:60]: nit: just reading through this, the name was a bit confusing at first.
Maybe we should get rid of the helper functions and simply have `declarativeconfig.CanModifyResource()` return an `errox.NotAuthorized` with appropriate error message, similar to what we have done with `sac.VerifyAuthZOK`?
This way we have the appropriate HTTP status code the same across datastore and it doesn't require each datastore to redefine the errox type.

File: central/role/datastore/datastore_impl.go
Comment (dhaus67) [inline central/role/datastore/datastore_impl.go:68]: So in case someone would use the `UpsertRole` for this datastore, they could only do it in case the role doesn't reference a declaratively created access scope / permission set? Don't think we want to make this constraint just now.

File: central/role/datastore/datastore_impl.go
Comment (dhaus67) [inline central/role/datastore/datastore_impl.go:133]: We should check the access scope's origin also when its a new one - which we currently wouldn't do.

File: central/role/datastore/datastore_impl.go
Comment (dhaus67) [inline central/role/datastore/datastore_impl.go:134]: Pretty sure we don't want to return here just yet.

**Review by dhaus67 — CHANGES_REQUESTED** [review]

File: central/declarativeconfig/manager_impl.go
Comment (ivan-degtiarenko) [reply central/declarativeconfig/manager_impl.go:205]: I think it's relatively easy to hardcode or parameterize order of reconciliation - for now I decided to go with the first approach by introducing list of types in order of reconciliation.

I think what's good about this approach is that by avoiding extra errors we shorten the time for all declarative resources to be reconciled(because if we fail to create a role on the first run, it will only show up in the system on the second run)

Comment (ivan-degtiarenko) [reply central/declarativeconfig/manager_impl.go:205]: That's not so useful considering 3/5 of these functions have the same underlying datastore which has the same lock for access scopes, permission sets and roles

File: central/declarativeconfig/manager_impl.go
Comment (ivan-degtiarenko) [reply central/declarativeconfig/manager_impl.go:192]: Do you mean datastore tests?

File: central/group/datastore/datastore_impl.go
Comment (ivan-degtiarenko) [reply central/group/datastore/datastore_impl.go:51]: I think that depends on whether we want to treat `Upsert` as the function that can be used outside the declarative context or not.

If it were to be used outside the declarative context, then we shouldn't require ID. If we only use this function within declarative context, we know for sure group will have deterministic ID set

File: central/role/datastore/datastore_impl.go
Comment (ivan-degtiarenko) [reply central/role/datastore/datastore_impl.go:133]: That's a very bad Ctrl-C Ctrl-V 😞

File: central/role/datastore/datastore_impl.go
Comment (ivan-degtiarenko) [reply central/role/datastore/datastore_impl.go:68]: If I understand correctly, we want to allow imperative reference declarative, but not other way around.
Do you want to introduce this constraint within this PR, or move it to a separate ticket(this one, for example: https://issues.redhat.com/browse/ROX-14999)?

File: central/group/datastore/datastore_impl.go
Comment (ivan-degtiarenko) [reply central/group/datastore/datastore_impl.go:60]: I think what's good about helper functions is that they allow to:
- specify the type of the resource
- specify the id of the resource
The second one is especially significant since we can't generalize it - the id for the group is accessed via `GetProps.GetId()` while for the rest of resources it's just `GetId`

Regarding helper function names: I agree they are far from ideal :) WDYT about just `verifyGroupOrigin` as an alternative?

File: central/declarativeconfig/updater/access_scope_updater.go
Comment (dhaus67) [inline central/declarativeconfig/updater/access_scope_updater.go:30]:
```suggestion
		return errox.InvariantViolation.Newf("wrong type passed to access scope updater: %T", accessScope)
```

File: central/declarativeconfig/updater/auth_provider_updater.go
Comment (dhaus67) [inline central/declarativeconfig/updater/auth_provider_updater.go:33]:
```suggestion
		return errox.InvariantViolation.Newf("wrong type passed to auth provider updater: %T", authProvider)
```

File: central/declarativeconfig/updater/auth_provider_updater.go
Comment (dhaus67) [inline central/declarativeconfig/updater/auth_provider_updater.go:14]: no reason to export those, right?
```suggestion
type authProviderUpdater struct {
```

File: central/declarativeconfig/updater/access_scope_updater.go
Comment (dhaus67) [inline central/declarativeconfig/updater/access_scope_updater.go:13]: no reason to export those, right?
```suggestion
type accessScopeUpdater struct {
```

File: central/declarativeconfig/updater/group_updater.go
Comment (dhaus67) [inline central/declarativeconfig/updater/group_updater.go:13]: no reason to export those, right?
```suggestion
type groupUpdater struct {
```

File: central/declarativeconfig/updater/group_updater.go
Comment (dhaus67) [inline central/declarativeconfig/updater/group_updater.go:30]:
```suggestion
		return errox.InvariantViolation.Newf("wrong type passed to group updater: %T", group)
```

File: central/declarativeconfig/updater/permission_set_updater.go
Comment (dhaus67) [inline central/declarativeconfig/updater/permission_set_updater.go:13]: no reason to export those, right?
```suggestion
type permissionSetUpdater struct {
```

File: central/declarativeconfig/updater/permission_set_updater.go
Comment (dhaus67) [inline central/declarativeconfig/updater/permission_set_updater.go:30]:
```suggestion
		return errox.InvariantViolation.Newf("wrong type passed to permission set updater: %T", permissionSet)
```

File: central/declarativeconfig/updater/role_updater.go
Comment (dhaus67) [inline central/declarativeconfig/updater/role_updater.go:13]: no reason to export those, right?
```suggestion
type roleUpdater struct {
```

File: central/declarativeconfig/updater/role_updater.go
Comment (dhaus67) [inline central/declarativeconfig/updater/role_updater.go:30]:
```suggestion
		return errox.InvariantViolation.Newf("wrong type passed to role updater: %T", role)
```

File: central/declarativeconfig/singleton.go
Comment (dhaus67) [inline central/declarativeconfig/singleton.go:42]: How about moving this map initialization call to the `updater` package, in a `DefaultResourceUpdater()` function or something like that?

It's more of a nit, but right now the manager has imports it shouldn't necessarily require / know (talking about the datastore dependencies).

File: central/group/datastore/datastore_impl.go
Comment (dhaus67) [inline central/group/datastore/datastore_impl.go:35]: Why not `sac.VerifyAuthZOK(accessSAC.WriteAllowed(ctx))`?

File: central/role/datastore/datastore_impl.go
Comment (dhaus67) [reply central/role/datastore/datastore_impl.go:68]: I guess we can move it to the separate ticket, fine by me.

File: central/declarativeconfig/updater/updater.go
Comment (dhaus67) [inline central/declarativeconfig/updater/updater.go:12]: super-nit: I wonder whether just calling it `ResourceHandler` would be more clear, given that it will handle both create / delete, but 🤷 
I don't have too strong of an opinion to block on it, so feel free to ignore that one.

File: central/declarativeconfig/manager_impl.go
Comment (dhaus67) [reply central/declarativeconfig/manager_impl.go:186]: I guess both? It'd be nice to test the flow once, specifically a) the `reconcileTransformedMessages` one regarding transforming messages and calling the updaters (with mocked updaters), ensuring we handle errors correctly as well as call the expected updaters and b) yeah datastore tests as well (seems like for role, permission set, access scope we already have some, but would be great to have one for group as well).

**Review by dhaus67 — COMMENTED** [review]
Minor comments as well as one comment regarding adding tests, otherwise LGTM 🚢 

Before merging this one, do we want to do some manual testing for this one?
While we are going to add automated once the E2E flow is done (so soon™) I think it'd be worthwhile testing the creation flow?

File: central/declarativeconfig/manager_impl.go
Comment (dhaus67) [reply central/declarativeconfig/manager_impl.go]: I agree it's not necessarily required for now.

I'm fine with leaving it for now given the limited number of resources, but in the future we might consider not depending on a specific order if the number of resources grows.

Comment (dhaus67) [reply central/declarativeconfig/manager_impl.go]: Yeah, as said more or less optional for the current time being, and only once we have a higher scale.

File: central/group/datastore/datastore_impl.go
Comment (dhaus67) [reply central/group/datastore/datastore_impl.go]: I suppose there's nothing stopping anyone using the datastore's generic interface to use the `Upsert` function currently.

The cleanest solution would be IMO to make it generic, and also usable within the non-declarative context. I don't see why a generic interface has a specific function only usable in a specific context - if we make it part of the generic interface, it should serve a generic purpose.

File: central/group/datastore/datastore_impl.go
Comment (dhaus67) [reply central/group/datastore/datastore_impl.go]: `verifyGroupOrigin` sounds good to me.

Yeah, I can agree with the type of the resource one, the `ID` one isn't really a reason for me since we could extend the `ResourceWithTraits` interface to include `GetId()`, and since we anyways pass `groups.GetProps()` to the `CanModifyResource()` function, it wouldn't really matter.

But, let's keep the helper for now.

File: central/declarativeconfig/manager_impl.go
Comment (ivan-degtiarenko) [reply central/declarativeconfig/manager_impl.go:202]: Since the customer is not in control of which proto types can be YAML configurations be transformed to and which proto types can be updated, the lack of corresponding updater is an error of us writing incorrect code. I think it is basically an incorrect state of the system, and perhaps we should stop entire `manager_impl` if we see such a type. That's an argument against sending this error down the `ProcessError` pipe.

The argument for sending this error is that `System Health` currently is the only page where this error will be visible to customers. Perhaps the flow should be:
1. If we encounter a such error, we send it to `ProcessError` with the message `FATAL: Manager cannot process proto type {proto type}. Processing declarative configurations will stop now. Please contact support to investigate the issue`
2. After that we immediately stop `manager_impl` from running

File: central/declarativeconfig/updater/updater.go
Comment (ivan-degtiarenko) [reply central/declarativeconfig/updater/updater.go:12]: I guess what prevented me from going for `ResourceHandler` is the fact that we already have a concept of handlers in declarative configuration, as well as a field in `manager_impl` called `transformedMessagesByHandler`

**Review by dhaus67 — APPROVED** [review]
LGTM modulo the few comments for proper comments in `central/declarativeconfig/types/accepted_types.go` 🚢

**Pull request merged**
